### ISSUE TITLE
need a way to get engine filed of EVP_PKEY

### ISSUE THREAD

**Issue by zhaozg** [openssl/openssl#1779]
need a way to get engine filed of EVP_PKEY, just like EVP_PKEY_get_engine(), version 1.1.x or master?

https://github.com/openssl/openssl/blob/6f0ac0e2f27d9240516edb9a23b7863e7ad02898/crypto/include/internal/evp_int.h

```c
struct evp_pkey_ctx_st {
    /* Method associated with this operation */
    const EVP_PKEY_METHOD *pmeth;
    /* Engine that implements this method or NULL if builtin */
    ENGINE *engine;
    /* Key: may be NULL */
    EVP_PKEY *pkey;
    /* Peer key for key agreement, may be NULL */
    EVP_PKEY *peerkey;
    /* Actual operation */
    int operation;
    /* Algorithm specific data */
    void *data;
    /* Application specific data */
    void *app_data;
    /* Keygen callback */
    EVP_PKEY_gen_cb *pkey_gencb;
    /* implementation specific keygen data */
    int *keygen_info;
    int keygen_info_count;
} /* EVP_PKEY_CTX */ ;
```

**Comment by evakili** [comment]
Also is there a way to get type field?

**Comment by snhenson** [comment]
EVP_PKEY_id() returns the type field.

**Comment by jon-oracle** [comment]
Simplest way would be like EVP_PKEY_id():

ENGINE* EVP_PKEY_engine(const EVP_PKEY *pkey)
{
    return pkey->engine;
}

But maybe this needs reference management with get0/get1 APIs?

EVP_PKEY_get0_engine()
EVP_PKEY_get1_engine()

**Comment by snhenson** [comment]
I'm curious: why do you need to get that field?

**Comment by petrovr** [comment]
Hi Stephen,
Stephen Henson wrote:
> I'm curious: why do you need to get that field?
>
I'm not sure why George ask for EVP_PKEY... function.

For RSA and DSA there is method like RSA_get0_engine and 
DSA_get0_engine, so in few steps
engine could be obtained :
...
  switch (EVP_PKEY_base_id(pkey)) {
     case EVP_PKEY_RSA: { RSA *rsa = EVP_PKEY_get0_RSA(pkey); e = 
RSA_get0_engine(rsa); ....
...

OpenSSL 1.1+ lack equivalent function for EC keys and there is no way to 
assess engine.

For instance in e_nss I store on engine as extra data some information 
like debug level and etc.

As work around I use global variable.

**Comment by jon-oracle** [comment]
Yes only EC keys are missing this method. Perhaps just adding an EC_KEY_get0_engine() for completeness is all that's required?

Also I've noticed that there's an ENGINE* inside evp_pkey_st as well as inside each 'union { dsa/rsa/dh/ec } pkey'. Is this duplication necessary?

**Comment by levitte** [comment]
> Also I've noticed that there's an ENGINE* inside evp_pkey_st as well as inside each 'union { dsa/rsa/dh/ec } pkey'. Is this duplication necessary?

Depends on what functionality you use.  If the caller ends up using the EVP interface, chances are that the EVP_PKEY engine pointer is enough.  However, if the caller uses the raw algo functions on the raw keys (i.e. for example pick out the RSA* from the EVP_PKEY and uses the RSA functions), then the raw key level engine pointer will be used.

We try to encourage people to use the EVP interface.  I assume the transition is going on... slowly.

**Comment by levitte** [comment]
@petrovr, I've an answer for you: #2960
